exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java

```diff
@@ -926,16 +944,22 @@ public GroupScan applyLimit(long maxRecords) {
       fileNames.add(rowGroupInfo.getPath());
     }
 
-    if (fileNames.size() == fileSet.size() ) {
+    // If there is no change in fileSet and maxRecords is >= batchSize, no need to create new groupScan.
+    if (fileNames.size() == fileSet.size() && (maxRecords >= batchSize) ) {
       // There is no reduction of rowGroups. Return the original groupScan.
       logger.debug("applyLimit() does not apply!");
       return null;
     }
 
+    // If limit maxRecords is less than batchSize, update batchSize to the limit size.
+    if (maxRecords < batchSize) {
+      batchSize = (int) maxRecords;
```

Comment:
I tried to avoid this casting.  To fix this, have to change code in bunch of places.  I was trying to keep code changes minimal so no new regressions are introduced.